### PR TITLE
strv: add new strv_extend_many() helper

### PR DESCRIPTION
**Pull request by poettering**
In vmspawn.c specifically there are just too many `strv_extend_strv(l, STRV_MAKE(…), false)` calls, that are ugly, cumbersome and also slow.

Let's turn this into `strv_extend_many(l, …)`, which is clearer to read.

Just some refactoring/clean-ups.

### CODE REVIEW

File: src/basic/strv.c [hunk 1]
```diff
@@ -511,23 +511,23 @@ int strv_insert(char ***l, size_t position, char *value) {
         n = strv_length(*l);
```
Comment (YHNdnzj) [inline src/basic/strv.c:513]: While at it, please add `assert(l)`

File: src/basic/strv.c [hunk 2]
```diff
@@ -588,6 +588,64 @@ int strv_extend_with_size(char ***l, size_t *n, const char *value) {
         return strv_consume_with_size(l, n, v);
 }
 
+int strv_extend_many_internal(char ***l, const char *value, ...) {
+        va_list ap;
+        size_t n, m;
+        int r;
+
```
Comment (YHNdnzj) [inline src/basic/strv.c:597]: assert(l)

File: src/vmspawn/vmspawn.c [hunk 3]
```diff
@@ -516,22 +516,22 @@ static int run_virtual_machine(void) {
...
-                ),  /* filter_duplicates= */ false);
+                r = strv_extend_many(
+                                &cmdline,
+                                "-nographic",
+                                "-nodefaults",
+                                "-chardev", "stdio,mux=on,id=console,signal=off",
+                                "-serial", "chardev:console",
+                                "-mon", "console");
                 if (r < 0)
                         return log_oom();
```
Comment (YHNdnzj) [inline src/vmspawn/vmspawn.c]: The error handling can be merged with that of the other branch, but doesn't matter.

File: src/basic/strv.c [hunk 4]
```diff
@@ -588,40 +588,62 @@ int strv_extend_with_size(char ***l, size_t *n, const char *value) {
...
-        if (!value)
-                return 0;
+        r = 0;
+        va_start(ap, value);
+        for (const char *s = value; s != POINTER_MAX; s = va_arg(ap, const char*)) {
+                if (!s)
+                        continue;
 
-        n = strv_length(*l);
+                if (m > SIZE_MAX-1) { /* overflow */
```
Comment (YHNdnzj) [inline src/basic/strv.c:608]: Just compare with `SIZE_MAX - 2` here?

File: src/basic/strv.c [hunk 5]
```diff
@@ -588,40 +588,62 @@ int strv_extend_with_size(char ***l, size_t *n, const char *value) {
...
+                m++;
+        }
+        va_end(ap);
 
-        /* Increase and overflow check. */
-        m = n + 2;
-        if (m < n)
+        if (r < 0)
+                return r;
+        if (m > SIZE_MAX-1)
```
Comment (YHNdnzj) [inline src/basic/strv.c:618]: (And drop this)

File: src/basic/strv.c [hunk 6]
```diff
@@ -588,40 +588,62 @@ int strv_extend_with_size(char ***l, size_t *n, const char *value) {
...
+        char **c = reallocarray(*l, GREEDY_ALLOC_ROUND_UP(m+1), sizeof(char*));
+        if (!c)
                 return -ENOMEM;
+        *l = c;
 
-        c = reallocarray(*l, m, sizeof(char*));
-        if (!c) {
-                free(v);
-                return -ENOMEM;
+        r = 0;
```
Comment (YHNdnzj) [inline src/basic/strv.c:626]: Hmm, is this actually needed?

File: src/basic/strv.c [hunk 7]
```diff
@@ -588,40 +588,62 @@ int strv_extend_with_size(char ***l, size_t *n, const char *value) {
...
+                /* rollback on error */
+                for (size_t j = n; j < i; j++)
+                        c[j] = mfree(c[j]);
+                return r;
+        }
 
-        *l = c;
-        return 0;
+        c[i] = NULL;
+        return r;
```
Comment (YHNdnzj) [inline src/basic/strv.c]: Just return 0 here instead?

File: src/basic/strv.c (hunk 4 above)
Comment (poettering) [reply src/basic/strv.c:608]: well, this would not suffice, because if the number of strings to add is zero we'd bypass the check entirely then.

File: src/basic/strv.c (hunk 6 above)
Comment (poettering) [reply src/basic/strv.c:626]: well, i figured it might be good to allow invocation with no arguments. or all arguments being null.

**Comment by poettering** [comment]
Force pushed new version now, with comments addressed (though see discussion).

I also added some more tests that ensure strv_extend_many() works nicely with a NULL argument list...

File: src/basic/strv.c (hunk 6 above)
Comment (YHNdnzj) [reply src/basic/strv.c:626]: (Still, r only needs to be initialized once. But doesn't matter)

**Review by YHNdnzj — APPROVED** [review]
